File: solutions/typescript/0082.ts
/**
 * 82. 删除排序链表中的重复元素 II
 * 给定已排序链表头，删除所有出现重复的数字节点，只保留出现一次的数字。
 * 哨兵 + 双指针：prev 指向最后一个确认唯一的节点；
 * 若 curr.val === curr.next.val 则一路跳过所有该值节点，prev.next = curr；
 * 否则 prev 前进一位。
 *
 * 注：ListNode 类型由 LeetCode 平台注入，本文件用 untyped 写法以避免与
 * 本地测试或平台模板里的同名类冲突。
 */

interface LNode { val: number; next: LNode | null }

function deleteDuplicates(head: LNode | null): LNode | null {
    const sentinel: LNode = { val: 0, next: head };
    let prev: LNode = sentinel;
    let curr: LNode | null = head;
    while (curr) {
        if (curr.next && curr.val === curr.next.val) {
            const dup = curr.val;
            while (curr && curr.val === dup) {
                curr = curr.next;
            }
            prev.next = curr;
        } else {
            prev = curr;
            curr = curr.next;
        }
    }
    return sentinel.next;
}

export default deleteDuplicates;

File: solutions/rust/0082.rs
/// 82. 删除排序链表中的重复元素 II
/// 删除所有出现重复的数字节点，只保留出现一次的。
/// 由于 Rust Box 链表难以同时维护 prev 引用和修改 prev.next，
/// 这里走"提取值→过滤→重建"路线：单次扫描利用链表已排序的特性，
/// 判断 vals[i] 是否与左右邻居相等，相等则丢弃。
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//     pub val: i32,
//     pub next: Option<Box<ListNode>>,
// }
//
// impl ListNode {
//     #[inline]
//     fn new(val: i32) -> Self {
//         ListNode { val, next: None }
//     }
// }

pub fn delete_duplicates(head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
    // 1. 提取所有值
    let mut vals: Vec<i32> = Vec::new();
    let mut curr = head;
    while let Some(node) = curr {
        vals.push(node.val);
        curr = node.next;
    }

    // 2. 链表已排序，vals[i] 需保留当且仅当它不等于左右邻居
    let n = vals.len();
    let mut keep: Vec<i32> = Vec::with_capacity(n);
    for i in 0..n {
        let prev_dup = i > 0 && vals[i - 1] == vals[i];
        let next_dup = i + 1 < n && vals[i + 1] == vals[i];
        if !prev_dup && !next_dup {
            keep.push(vals[i]);
        }
    }

    // 3. 重建链表
    let mut new_head: Option<Box<ListNode>> = None;
    for &v in keep.iter().rev() {
        new_head = Some(Box::new(ListNode { val: v, next: new_head }));
    }
    new_head
}

impl Solution {
    pub fn delete_duplicates(head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        delete_duplicates(head)
    }
}